src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -97,8 +100,20 @@ class Executor {
                 : Status(StatusCode(reply.status().code()), reply.status().message()); \
         callback(status, reply);                                                       \
         delete executor;                                                               \
+      } else if (gcs_is_down_) {                                                       \
+        pending_requests_.emplace_back(executor);                                      \
+        if (pending_requests_.size() >                                                 \
+            ::RayConfig::instance().gcs_grpc_max_request_queued()) {                   \
+          RAY_LOG(WARNING) << "Pending queue for failed GCS request has reached the "  \
+                           << "limit. Blocking the current thread until GCS is back";  \
+          while (gcs_is_down_ && !shutdown_) {                                         \
+            CheckChannelStatus();                                                      \
+            std::this_thread::sleep_for(std::chrono::milliseconds(                     \
+                ::RayConfig::instance()                                                \
+                    .gcs_client_check_connection_status_interval_milliseconds()));     \
+          }                                                                            \
+        }                                                                              \
       } else {                                                                         \
```

Comment:
Why do we need to handle `gcs_is_down_ = true` vs `false` differently? It seem even when `gcs_is_down_ = false`, GCS can be down in reality but not yet discovered by periodic check status. Retry immediately with `executor->Retry()` should work, but it seems more consistent to handle both situations by checking GCS status then retry if connected. Or I may be missing something.

Reply:
It's right. Immediately retry should work.

But the concern here is 1) save CPU since always retry immediately definitely will overload the thread. 2) for memory safety. if GCS is down and there are a lot of requests coming, we'd like to make sure OOM won't happen here. Or at least we'll have some way to control this.

The previous implementation will trigger a blocking reconnection so it's not an issue there.

Reply:
Good point here. I think I'll change the behavior to:

- if rpc failed, I'll check channel connection first
- and if gcs is not down, I'll retry this.
- otherwise queue this